### ISSUE TITLE
Large gifs take a long time to appear

### ISSUE THREAD

**Issue by crowbarsolutions** [koral--/android-gif-drawable#352]
So I have a ~150MB GIF file on my device.

I create a FileInputStream from that file, and wrap it in a BufferedInputStream

```
File myFile = new File("myFile.gif");
FileInputStream fileStream = new FileInputStream(myFile);
BufferedInputStream bufferedStream = new BufferedInputStream(fileStream);
```

Next, I create a GifDrawable from the BufferedInputStream I created, and add sampling.

```
GifDrawableBuilder builder = new GifDrawableBuilder();
builder.sampleSize(4);
builder.from(bufferedStream);
GifDrawable gifDrawable = builder.build();
```

I then take the GifDrawable I created and set it as the image for a GifImageView that I have defined in my layout xml file.

```
gifView = (GifImageView) findViewById(R.id.mygifview);
gifView.setImageDrawable(gifDrawable);
```

The actual execution of the code happens quickly, but it can take over a minute until the GIF appears on my screen. In the meantime logcat is spamming:

> Starting a blocking GC Alloc
> Alloc concurrent mark sweep GC freed 3(96B) AllocSpace objects, 0(0B) LOS objects, 15% free, 87MB/103MB, paused 1.763ms total 31.407ms
> Forcing collection of SoftReferences for 128MB allocation
> Starting a blocking GC Alloc
> Alloc concurrent mark sweep GC freed 3(96B) AllocSpace objects, 0(0B) LOS objects, 15% free, 87MB/103MB, paused 1.717ms total 31.387ms
> Throwing OutOfMemoryError "Failed to allocate a 134217740 byte allocation with 16776992 free bytes and 104MB until OOM"
> Starting a blocking GC Alloc
> Starting a blocking GC Alloc
> etc...

After a minute goes by, logcat stops spamming and the GIF plays smoothly with no problems. I experimented with different buffer sizes for the BufferedInputStream as well as different sampling sizes and it didn't make much of a difference.

Any advice on how I can have the GIF file load more quickly?

**Edit:** With some more experimentation it seems that sampling has little to no effect on the loading time. However, when creating the BufferedInputStream, it seems some improvement was found when specifying the buffer size:

```
int bufferSize = (int) myFile.length;
new BufferedInputStream(fileStream, bufferSize);
```

I found that setting the buffer size equal to the byte length of the GIF file resulted in a big improvement. My 150MB file would load in roughly 14 seconds. When the buffer size was set to anything less than the size of the GIF file, performance would greatly suffer.

Obviously this is an issue because you can't set the buffer size to be equal to the GIF file size when you have a 150MB file. You end up with OutOfMemoryError being thrown when you try to instantiate the BufferedInputStream. What are your thoughts?

**Comment by koral--** [comment]
For such large GIFs the best option is to create `GifDrawable` from `File` or path (`String`) like that: 
```java
new GifDrawable("/path/myFile.gif");
//or
File myFile = new File("myFile.gif");
new GifDrawable(myFile);
```
I've just checked ~290 MB GIF (1280x720, ~1000 frames) on Nexus 5X, and it takes ~3 seconds to create `GifDrawable` from file path vs more than 10 minutes from `InputStream` (I haven't waited till the end).

Reading from `InputStream` is much slower in case of large files because there is a 256 B buffer used under the hood.

Setting sample size doesn't affect loading time significantly in case of long files but not extremely large frames. No matter what is the sample size entire file needs to be read at the loading time to know number of frames, the largest frame size etc. After that, frame buffer is allocated - its size depends on sample size.

**Comment by crowbarsolutions** [comment]
My GIF files are encrypted, at the moment I use a custom FileInputStream to decrypt the bytes as they are read. I can't pass in just the path of the file as it would not appear as a valid GIF file. 

Besides creating a temporary decrypted copy of the GIF file and passing that in directly, is there another way I could pass it to the GifDrawable class, in place, such that I can perform operations on the bytes as they are read?

Is it perhaps possible to change the 256B buffer so that it uses the same buffer size as the BufferedInputStream?

**Comment by koral--** [comment]
Solution other than creating temporary unencrypted file is to buffer entire file in `byte[]` or direct `ByteBuffer`.
It should be possible to enlarge that internal buffer. I'll try to do that, but can't promise any ETA. However pull requests is always welcome.
In the meantime I'll keep this issue open.

**Comment by Innani** [comment]
So I have 2MB gif file ~100 frames in drawable resource.
It take about 10 minutes to start that gif in the first time I launch the apps and when I clear the cache.
How can I resolve this problem?
Thank you.

**Comment by koral--** [comment]
Buffer has now 8KB, available in `1.2.4-SNAPSHOT`.
